src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldNotContainUnderscores.cs

```diff
@@ -31,78 +36,208 @@ public abstract class IdentifiersShouldNotContainUnderscoresAnalyzer : Diagnosti
+                                    }
+
+                                    symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(TypeRule, symbol.ToDisplayString()));
+                                    return;
+                                }
+
+                            case SymbolKind.Field:
+                                {
+                                    var fieldSymbol = symbol as IFieldSymbol;
+                                    if (symbol.IsPublic() && (fieldSymbol.IsConst || (fieldSymbol.IsStatic && fieldSymbol.IsReadOnly)))
+                                    {
+                                        symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(MemberRule, symbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)));
+                                        return;
+                                    }
+
+                                    return;
+                                }
+
+                            default:
+                                {
+                                    if (IsInvalidSymbol(symbol))
+                                    {
+                                        return;
+                                    }
+
+                                    symbolAnalysisContext.ReportDiagnostic(symbol.CreateDiagnostic(MemberRule, symbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)));
+                                    return;
+                                }
+                        }
+                    }
+                },
+                SymbolKind.Namespace, // Namespace
+                SymbolKind.NamedType, //Type
+                SymbolKind.Method, SymbolKind.Property, SymbolKind.Field, SymbolKind.Event // Members
+                );
+
+                compilationStartAnalysisContext.RegisterSyntaxNodeAction(syntaxNodeAnalysisContext =>
+                {
+                    AnalyzeSyntaxNode(syntaxNodeAnalysisContext);
+                }, SyntaxKinds);
+            });
+
+            analysisContext.RegisterCompilationAction(compilationAnalysisContext =>
+            {
+                var compilation = compilationAnalysisContext.Compilation;
+                if (ContainsUnderScore(compilation.AssemblyName))
+                {
+                    compilationAnalysisContext.ReportDiagnostic(compilation.Assembly.CreateDiagnostic(AssemblyRule, compilation.AssemblyName));
+                }
+            });
+        }
+
+        private bool IsInvalidSymbol(ISymbol symbol)
+        {
+            return (!(symbol.GetResultantVisibility() == SymbolVisibility.Public && !symbol.IsOverride)) ||
+                symbol.IsAccessorMethod() || symbol.IsInterfaceMemberImplementation();
+        }
+
+        protected void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext)
+        {
+            var symbol = syntaxNodeAnalysisContext.SemanticModel.GetDeclaredSymbol(syntaxNodeAnalysisContext.Node);
```

Comment:
@srivatsn and @heejaechang have both suggested at this point to avoid the syntax node analysis so making the change to avoid the syntax nodes